spring-kafka/src/main/java/org/springframework/kafka/support/serializer/DelegatingByTopicDeserializer.java

```diff
@@ -0,0 +1,229 @@
+			if (deserializer instanceof Deserializer) {
+				this.delegates.put(pattern, (Deserializer<?>) deserializer);
+				((Deserializer<?>) deserializer).configure(configs, isKey);
+			}
+			else if (deserializer instanceof Class) {
+				instantiateAndConfigure(configs, isKey, this.delegates, pattern, (Class<?>) deserializer);
+			}
+			else if (deserializer instanceof String) {
+				createInstanceAndConfigure(configs, isKey, this.delegates, pattern, (String) deserializer);
+			}
+			else {
+				throw new IllegalStateException(configKey
+						+ " map entries must be Serializers or class names, not " + value.getClass());
+			}
+		});
+	}
+
+	private String configKey() {
+		return this.forKeys
+				? DelegatingByTopicSerializer.KEY_SERIALIZATION_TOPIC_CONFIG
+				: DelegatingByTopicSerializer.VALUE_SERIALIZATION_TOPIC_CONFIG;
+	}
+
+	protected Map<Pattern, Deserializer<?>> createDelegates(String mappings, Map<String, ?> configs,
+			boolean isKey) {
+
+		Map<Pattern, Deserializer<?>> delegateMap = new HashMap<>();
+		String[] array = StringUtils.commaDelimitedListToStringArray(mappings);
+		for (String entry : array) {
+			String[] split = entry.split(":");
+			Assert.isTrue(split.length == 2, "Each comma-delimited entry entry must have exactly one ':'");
+			createInstanceAndConfigure(configs, isKey, delegateMap, obtainPattern(split[0]), split[1]);
+		}
+		return delegateMap;
+	}
+
+	protected static void createInstanceAndConfigure(Map<String, ?> configs, boolean isKey,
+			Map<Pattern, Deserializer<?>> delegateMap, Pattern pattern, String className) {
+
+		try {
+			Class<?> clazz = ClassUtils.forName(className.trim(), ClassUtils.getDefaultClassLoader());
+			instantiateAndConfigure(configs, isKey, delegateMap, pattern, clazz);
+		}
+		catch (ClassNotFoundException | LinkageError e) {
+			throw new IllegalArgumentException(e);
+		}
+	}
+
+	private Pattern obtainPattern(Object key) {
+		if (key instanceof Pattern) {
+			return (Pattern) key;
+		}
+		else if (key instanceof String) {
+			if (this.cased) {
+				return Pattern.compile(((String) key).trim());
+			}
+			else {
+				return Pattern.compile(((String) key).trim(), Pattern.CASE_INSENSITIVE);
+			}
+		}
+		else {
+			throw new IllegalStateException("Map key must be a Pattern or a String, not a " + key.getClass());
+		}
+	}
+
+	protected static void instantiateAndConfigure(Map<String, ?> configs, boolean isKey,
+			Map<Pattern, Deserializer<?>> delegateMap, Pattern pattern, Class<?> clazz) {
+
+		try {
+			Deserializer<?> delegate = (Deserializer<?>) clazz.getDeclaredConstructor().newInstance();
+			delegate.configure(configs, isKey);
+			delegateMap.put(pattern, delegate);
+		}
+		catch (Exception e) {
+			throw new IllegalArgumentException(e);
+		}
+	}
+
+	public void addDelegate(Pattern pattern, Deserializer<?>  deserializer) {
+		this.delegates.put(pattern, deserializer);
+	}
+
+	@Nullable
+	public Deserializer<?> removeDelegate(String selector) {
+		return this.delegates.remove(selector);
```

Comment:
How does it work if we put by `Pattern`, but remove by `String`?

Reply:
oops (copy/paste)